src/python/pants/backend/java/classpath.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from dataclasses import dataclass
+from typing import Callable, Iterator
+
+from pants.backend.java.compile.javac import CompiledClassfiles, CompileJavaSourceRequest
+from pants.engine.addresses import Addresses
+from pants.engine.fs import AddPrefix, Digest, MergeDigests, Snapshot
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    CoarsenedTargets,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.jvm.resolve.coursier_fetch import (
+    ArtifactRequirements,
+    Coordinate,
+    CoursierLockfileForTargetRequest,
+    CoursierResolvedLockfile,
+    MaterializedClasspath,
+    MaterializedClasspathRequest,
+)
+
+_USERCP_RELPATH = "__usercp"
+
+
+@dataclass(frozen=True)
+class Classpath:
+    """A transitive classpath which is sufficient to launch the target(s) it was generated for.
+
+    This classpath is guaranteed to contain only JAR files.
+    """
+
+    content: Snapshot
+
+    def classpath_args(self, prefix: str | None = None) -> Iterator[str]:
```

Comment:
This name/docstring is not strictly true: the argument to `java -cp` needs to be `:`-separated, this isn't. Maybe the method could be `classpath_filenames` and the docstring

> Return filenames that can be provided to tools that require this classpath